third_party/lua/luadebug.lua

```diff
@@ -413,13 +480,148 @@ local function cmd_eval(code)
+local function _bp_format_expected()
+    dbg_write_error("command expects argument in format filename:NN or " ..
+                    "filename+NN, where NN should be a positive number.")
+end
+
+-- check range to fit in boundaries {from, to}
+local function check_line_max(v, from, to)
+    if type(v) ~= 'number' then
+        dbg_write_error("numeric value expected as line number, " ..
+                        "but received %s", type(v))
+        return false
+    end
+    if v >= from and v <= to then
+        return true
+    end
+    dbg_write_error("line number %d is out of allowed range [%d, %d]",
+                    v, from, to)
+    return false
+end
+
+
+local function cmd_add_breakpoint(bps)
+    assert(bps)
+    local fullfile, line = parse_bp(bps)
+    if not fullfile then
+        _bp_format_expected()
+        return false
+    end
+    if not check_line_max(line, 1, LJ_MAX_LINE) then
+        return false
+    end
+    local info = debug.getinfo(stack_inspect_offset + CMD_STACK_LEVEL, "S")
+    local debuggee = info.source
+    local file = #fullfile > 0 and fullfile or debuggee
+    local normfile = normalize_path(file)
+
+    -- save direct hash (for faster lookup): line -> file
+    if not breakpoints[line] then
+        breakpoints[line] = {}
+    end
+    breakpoints[line][normfile] = true
+    -- save reversed hash information: file -> [lines]
+    if not breakpoints[normfile] then
+        breakpoints[normfile] = {}
+    end
+    breakpoints[normfile][line] = true
+
+    dbg_writeln("Added breakpoint %s:%d.", color_blue(normfile), line)
+    where({source = file, currentline = line}, 0, true)
+    return false
+end
+
+local function _bpd_format_expected()
+    dbg_write_error(" command expects argument specifying breakpoint or" ..
+                    " * for all breakpoints.")
+end
+
+local function cmd_remove_breakpoint(bps)
+    assert(bps)
+    if bps == '*' then
+        breakpoints = {}
+        dbg_writeln("Removed all breakpoints.")
+        return false
+    end
+
+    local fullfile, line = parse_bp(bps)
+    if not fullfile then
+        _bpd_format_expected()
+        return false
+    end
+    if not check_line_max(line, 1, LJ_MAX_LINE) then
+        return false
+    end
+    local info = debug.getinfo(stack_inspect_offset + CMD_STACK_LEVEL, "S")
+    local debuggee = info.source
+    local file = #fullfile > 0 and fullfile or debuggee
+    local normfile = normalize_path(file)
+    local removed = false
+
+    if breakpoints[line] and breakpoints[line][normfile] then
+        breakpoints[line][normfile] = nil
+        -- direct and reverse hashes should be in sync
+        assert(breakpoints[normfile] and breakpoints[normfile][line])
+        breakpoints[normfile][line] = nil
+        removed = true
+    end
+    if removed then
+        dbg_writeln("Removed breakpoint %s:%d.", color_blue(normfile), line)
+    else
+        dbg_writeln("No breakpoint defined in %s:%d.", color_blue(normfile), line)
+    end
```

Comment:
Nit: `removed` is used just once and `dbg_writeln()` could be move above.

Reply:
You've lost me here entirely. Could not parse and get any meaning. 
What are you talking about here?

Reply:
```diff
diff --git a/third_party/lua/luadebug.lua b/third_party/lua/luadebug.lua
index ae8ab54ea..de7f012a6 100644
--- a/third_party/lua/luadebug.lua
+++ b/third_party/lua/luadebug.lua
@@ -578,16 +578,12 @@ local function cmd_remove_breakpoint(bps)
     local debuggee = info.source
     local file = #fullfile > 0 and fullfile or debuggee
     local normfile = normalize_path(file)
-    local removed = false
 
     if breakpoints[line] and breakpoints[line][normfile] then
         breakpoints[line][normfile] = nil
         -- direct and reverse hashes should be in sync
         assert(breakpoints[normfile] and breakpoints[normfile][line])
         breakpoints[normfile][line] = nil
-        removed = true
-    end
-    if removed then
         dbg_writeln("Removed breakpoint %s:%d.", color_blue(normfile), line)
     else
         dbg_writeln("No breakpoint defined in %s:%d.", color_blue(normfile), line)

    ```